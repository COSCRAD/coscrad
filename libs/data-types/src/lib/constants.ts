export const COSCRAD_DATA_TYPE_METADATA = '__coscrad-data-types__';

export const COSCRAD_UNION_MEMBER_METADATA = '__coscrad-union-member__';
