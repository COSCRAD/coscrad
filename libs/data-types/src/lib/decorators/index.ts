export * from './CompositeIdentifier';
export * from './Enum';
export * from './NestedDataType';
export * from './NonEmptyString';
export * from './NonNegativeFiniteNumber';
export * from './RawData';
export * from './URL';
export * from './UUID';
