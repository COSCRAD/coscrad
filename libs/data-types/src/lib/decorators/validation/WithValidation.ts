import { IsOptional } from '@coscrad/validation';
import { TypeDecoratorOptions } from '../types/TypeDecoratorOptions';

export default (
        validationDecorator: PropertyDecorator,
        { isOptional }: Pick<TypeDecoratorOptions, 'isOptional'>
    ): PropertyDecorator =>
    (target: Object, propertyKey: string | symbol) => {
        if (isOptional) {
            IsOptional()(target, propertyKey);
        }

        validationDecorator(target, propertyKey);
    };
