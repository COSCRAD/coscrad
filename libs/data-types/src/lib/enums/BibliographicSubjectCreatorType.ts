export enum BiblographicSubjectCreatorType {
    author = 'author',
    director = 'director',
    artist = 'artist',
}
