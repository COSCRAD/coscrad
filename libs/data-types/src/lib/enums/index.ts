export * from './BibliographicSubjectCreatorType';
export * from './CoscradEnum';
export * from './CoscradUserRole';
export * from './MIMEType';
