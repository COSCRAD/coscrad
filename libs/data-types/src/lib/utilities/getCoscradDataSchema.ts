import { ClassSchema } from '../types/ClassSchema';
import getCoscradDataSchemaFromPrototype from './getCoscradDataSchemaFromPrototype';

// eslint-disable-next-line
export default (TargetClass: Object): ClassSchema<Record<string, unknown>> =>
    // @ts-expect-error TODO: restrict argument to be a class
    getCoscradDataSchemaFromPrototype(TargetClass.prototype || {});
