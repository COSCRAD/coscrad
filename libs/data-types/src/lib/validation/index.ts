export * from './buildSimpleValidationFunction';
export * from './DiscriminatedUnionValidator';
