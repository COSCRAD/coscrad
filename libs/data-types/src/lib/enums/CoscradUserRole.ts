import { isEnum } from '@coscrad/validation';

export enum CoscradUserRole {
    superAdmin = 'superAdmin', // Technical Admin
    projectAdmin = 'projectAdmin', // Project admin- can read and write
    viewer = 'viewer', // Authenticated user- can read including non-public materials if they are in the read ACL
}

export const isCoscardUserRole = (input: unknown): input is CoscradUserRole =>
    isEnum(input, CoscradUserRole);
