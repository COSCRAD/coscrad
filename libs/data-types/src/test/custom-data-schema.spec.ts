import { buildSimpleValidationFunction } from '@coscrad/validation';
import 'reflect-metadata';
import { NonEmptyString, URL, UUID } from '../index';
import {
    DiscriminatedBy,
    Enum,
    ISBN,
    NestedDataType,
    NonNegativeFiniteNumber,
    PositiveInteger,
    RawDataObject,
    Union,
    Year,
} from '../lib/decorators';
import { BibliographicSubjectCreatorType, CoscradEnum, MIMEType } from '../lib/enums';
import { CoscradUserRole } from '../lib/enums/CoscradUserRole';
import getCoscradDataSchema from '../lib/utilities/getCoscradDataSchema';

const isOptional = true;

describe('NonEmptyString', () => {
    class Whatsit {
        @NonEmptyString({ isOptional })
        whatsitName = 'whatsit 1';

        @UUID()
        whatsitId = '25c5824f-6b4b-4341-bb60-3145d8109568';
    }

    @DiscriminatedBy('one')
    class ThingDataOne {
        type = 'one';

        @NonNegativeFiniteNumber()
        strength = 99.5;
    }

    @DiscriminatedBy('two')
    class ThingDataTwo {
        type = 'two';

        @PositiveInteger()
        rating = 5;
    }

    class Widget {
        @NonEmptyString()
        widgetName = 'Machine';

        @NonEmptyString({ isOptional })
        locationName = 'Back Red Room 12';

        // @IsString({ each: true })
        // @IsNotEmpty()
        @NonEmptyString({ isArray: true })
        aliases: ['super machine', 'widget king'];

        @UUID()
        id = '25c5824f-6b4b-4341-bb60-3145d8109568';

        @UUID({ isOptional })
        locationId = '25c5824f-6b4b-4341-bb60-3145d8109568';

        @URL()
        iconUrl = 'https://www.mylink.com/uploads/123.png';

        @URL({ isOptional })
        specSheetUrl = undefined;

        @NonNegativeFiniteNumber()
        width = 134.5;

        @NonNegativeFiniteNumber({ isOptional })
        averageRating = 3.5;

        @NestedDataType(Whatsit)
        primaryWhatsit = {};

        @NestedDataType(Whatsit, { isOptional })
        secondaryWhatsit = {};

        @RawDataObject()
        rawDataObject = { foo: 72 };

        @RawDataObject({ isOptional })
        optionalRawData = undefined;

        @Enum(CoscradEnum.MIMEType)
        mimeType = MIMEType.mp3;

        @Enum(CoscradEnum.CoscradUserRole)
        role = CoscradUserRole.viewer;

        @Enum(CoscradEnum.BibliographicSubjectCreatorType)
        creatorType = BibliographicSubjectCreatorType.artist;

        @Year()
        year = 1998;

        @Year({ isOptional })
        yearUploaded = 2002;

        @PositiveInteger()
        numberOfLikes = 22;

        @PositiveInteger({ isOptional })
        numberOfDownvotes = 2;

        @ISBN()
        requiredISBN = `978-3-16-148410-0`;

        @ISBN({ isOptional })
        optionalISBN = `979-3-16-148410-0`;

        @Union([ThingDataOne, ThingDataTwo], 'data.type')
        data: ThingDataOne | ThingDataTwo = {
            type: 'one',

            strength: 67.3,
        };

        constructor(dto: Widget) {
            Object.assign(this, dto);
        }
    }

    const validWidgetDTO: Widget = {
        widgetName: 'Machine',

        locationName: 'Back Red Room 12',

        aliases: ['super machine', 'widget king'],

        id: '25c5824f-6b4b-4341-bb60-3145d8109568',

        locationId: '25c5824f-6b4b-4341-bb60-3145d8109568',

        iconUrl: 'https://www.mylink.com/uploads/123.png',

        specSheetUrl: undefined,

        width: 134.5,

        averageRating: 3.5,

        primaryWhatsit: {},

        secondaryWhatsit: {},

        rawDataObject: { foo: 72 },

        optionalRawData: undefined,

        mimeType: MIMEType.mp3,

        role: CoscradUserRole.viewer,

        creatorType: BibliographicSubjectCreatorType.artist,

        year: 1945,

        yearUploaded: 1999,

        numberOfLikes: 2,

        numberOfDownvotes: 501,

        requiredISBN: `978-3-16-148410-0`,

        optionalISBN: `978-3-16-148410-0`,

        data: {
            type: 'one',
            strength: 85,
        },
    };

    it('should populate the appropriate metadata', () => {
        const actualMetadata = getCoscradDataSchema(Widget);

        expect(actualMetadata).toMatchSnapshot();
    });

    describe('the corresponding invariant validation', () => {
        describe('when the data is valid', () => {
            it('should return Valid', () => {
                const result = buildSimpleValidationFunction(Widget)(validWidgetDTO);

                expect(result).toStrictEqual([]);
            });
        });
    });

    /**
     * TODO [https://www.pivotaltracker.com/story/show/182578952]
     * test the invalid data cases
     *
     * It may be helpful to wait for the following story to be finished
     * [https://www.pivotaltracker.com/story/show/182217249]
     */
});
