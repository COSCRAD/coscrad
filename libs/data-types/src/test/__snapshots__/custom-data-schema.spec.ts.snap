// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NonEmptyString should populate the appropriate metadata 1`] = `
Object {
  "aliases": Object {
    "coscradDataType": "NON_EMPTY_STRING",
    "isArray": true,
    "isOptional": false,
  },
  "averageRating": Object {
    "coscradDataType": "NON_NEGATIVE_FINITE_NUMBER",
    "isArray": false,
    "isOptional": true,
  },
  "creatorType": Object {
    "coscradDataType": "ENUM",
    "enumLabel": "Creator Type",
    "enumName": "BIBLIOGRAPHIC_SUBJECT_CREATOR_TYPE",
    "isArray": false,
    "isOptional": false,
    "labelsAndValues": Array [
      Object {
        "label": "artist",
        "value": "artist",
      },
      Object {
        "label": "author",
        "value": "author",
      },
      Object {
        "label": "director",
        "value": "director",
      },
    ],
    "type": "ENUM",
  },
  "iconUrl": Object {
    "coscradDataType": "URL",
    "isArray": false,
    "isOptional": false,
  },
  "id": Object {
    "coscradDataType": "UUID",
    "isArray": false,
    "isOptional": false,
  },
  "locationId": Object {
    "coscradDataType": "UUID",
    "isArray": false,
    "isOptional": true,
  },
  "locationName": Object {
    "coscradDataType": "NON_EMPTY_STRING",
    "isArray": false,
    "isOptional": true,
  },
  "mimeType": Object {
    "coscradDataType": "ENUM",
    "enumLabel": "MIME Type",
    "enumName": "MIME_TYPE",
    "isArray": false,
    "isOptional": false,
    "labelsAndValues": Array [
      Object {
        "label": "mp3",
        "value": "audio/mpeg",
      },
      Object {
        "label": "mp4",
        "value": "video/mp4",
      },
    ],
    "type": "ENUM",
  },
  "numberOfDownvotes": Object {
    "isArray": false,
    "isOptional": true,
    "type": "POSITIVE_INTEGER",
  },
  "numberOfLikes": Object {
    "isArray": false,
    "isOptional": false,
    "type": "POSITIVE_INTEGER",
  },
  "optionalISBN": Object {
    "isArray": false,
    "isOptional": true,
    "type": "ISBN",
  },
  "optionalRawData": Object {
    "coscradDataType": "RAW_DATA",
    "isArray": false,
    "isOptional": true,
  },
  "primaryWhatsit": Object {
    "coscradDataType": Object {
      "whatsitId": Object {
        "coscradDataType": "UUID",
        "isArray": false,
        "isOptional": false,
      },
      "whatsitName": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "isArray": false,
        "isOptional": true,
      },
    },
    "isArray": false,
    "isOptional": false,
  },
  "rawDataObject": Object {
    "coscradDataType": "RAW_DATA",
    "isArray": false,
    "isOptional": false,
  },
  "requiredISBN": Object {
    "isArray": false,
    "isOptional": false,
    "type": "ISBN",
  },
  "role": Object {
    "coscradDataType": "ENUM",
    "enumLabel": "User Role",
    "enumName": "COSCRAD_USER_ROLE",
    "isArray": false,
    "isOptional": false,
    "labelsAndValues": Array [
      Object {
        "label": "admin",
        "value": "projectAdmin",
      },
      Object {
        "label": "viewer",
        "value": "viewer",
      },
      Object {
        "label": "COSCRAD admin",
        "value": "superAdmin",
      },
    ],
    "type": "ENUM",
  },
  "secondaryWhatsit": Object {
    "coscradDataType": Object {
      "whatsitId": Object {
        "coscradDataType": "UUID",
        "isArray": false,
        "isOptional": false,
      },
      "whatsitName": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "isArray": false,
        "isOptional": true,
      },
    },
    "isArray": false,
    "isOptional": true,
  },
  "specSheetUrl": Object {
    "coscradDataType": "URL",
    "isArray": false,
    "isOptional": true,
  },
  "widgetName": Object {
    "coscradDataType": "NON_EMPTY_STRING",
    "isArray": false,
    "isOptional": false,
  },
  "width": Object {
    "coscradDataType": "NON_NEGATIVE_FINITE_NUMBER",
    "isArray": false,
    "isOptional": false,
  },
  "year": Object {
    "isArray": false,
    "isOptional": false,
    "type": "YEAR",
  },
  "yearUploaded": Object {
    "isArray": false,
    "isOptional": true,
    "type": "YEAR",
  },
  "year": Object {
    "isArray": false,
    "isOptional": false,
    "type": "YEAR",
  },
  "yearUploaded": Object {
    "isArray": false,
    "isOptional": true,
    "type": "YEAR",
  },
}
`;
